#**Vektorisasi Part 2**
#**Nama : Alya Fauziyah**
#**NIM : 21110030**
#**Kelas : S1-SD02A**

In [4]:
!pip install PyPDF2

In [5]:
import nltk
nltk.download('punkt')

[nltk_data] Downloading package punkt to /root/nltk_data...
[nltk_data]   Package punkt is already up-to-date!


True

In [6]:
import PyPDF2
import pandas as pd
from gensim.models import Word2Vec, FastText
from nltk.tokenize import sent_tokenize, word_tokenize
from sklearn.feature_extraction.text import CountVectorizer

#**1. One hot encoding**

In [7]:
# Membaca teks dari file PDF
def extract_text_from_pdf(pdf_file):
    text = ""
    with open(pdf_file, 'rb') as file:
        reader = PyPDF2.PdfReader(file)
        for page_number in range(len(reader.pages)):
            text += reader.pages[page_number].extract_text()
    return text

# Mendapatkan teks dari file PDF
pdf_file = 'MALIN_KUNDANG.pdf'
data = extract_text_from_pdf(pdf_file)

In [8]:
data

'MALIN KUNDANG \nPada suatu waktu, hiduplah sebuah keluarga nelayan di \npesisir pantai wilayah Sumatra. Keluarga tersebut terdiri \ndari ayah, ibu dan seorang anak laki-laki yang diberi nama \nMalin Kundang. Karena kondisi keuangan keluarga yang \nmemprihatinkan, sang ayah memutuskan untuk mencari \nnafkah di negeri seberang dengan mengarungi lautan yang \nluas.\nMaka tinggallah si Malin dan ibunya di gubug mereka. Semingg u, dua minggu, sebulan, dua \nbulan bahkan sudah 1 tahun lebih lamanya, ayah Malin tidak  juga kembali ke kampung \nhalamannya. Sehingga ibunya harus menggantikan posisi ayah Malin untuk mencari nafkah. \nMalin termasuk anak yang cerdas tetapi sedikit nakal. Ia  sering mengejar ayam dan \nmemukulnya dengan sapu. Suatu hari ketika Malin sedang me ngejar ayam, ia tersandung \nbatu dan lengan kanannya luka terkena batu. Luka terse but menjadi berbekas dilengannya \ndan tidak bisa hilang. \nSetelah beranjak dewasa, Malin Kundang merasa kasihan dengan ibunya yang banting

In [13]:
token = sent_tokenize(data)
teks = ''.join(token)

corpus = word_tokenize(teks)
corpus

['MALIN',
 'KUNDANG',
 'Pada',
 'suatu',
 'waktu',
 ',',
 'hiduplah',
 'sebuah',
 'keluarga',
 'nelayan',
 'di',
 'pesisir',
 'pantai',
 'wilayah',
 'Sumatra.Keluarga',
 'tersebut',
 'terdiri',
 'dari',
 'ayah',
 ',',
 'ibu',
 'dan',
 'seorang',
 'anak',
 'laki-laki',
 'yang',
 'diberi',
 'nama',
 'Malin',
 'Kundang.Karena',
 'kondisi',
 'keuangan',
 'keluarga',
 'yang',
 'memprihatinkan',
 ',',
 'sang',
 'ayah',
 'memutuskan',
 'untuk',
 'mencari',
 'nafkah',
 'di',
 'negeri',
 'seberang',
 'dengan',
 'mengarungi',
 'lautan',
 'yang',
 'luas.Maka',
 'tinggallah',
 'si',
 'Malin',
 'dan',
 'ibunya',
 'di',
 'gubug',
 'mereka.Semingg',
 'u',
 ',',
 'dua',
 'minggu',
 ',',
 'sebulan',
 ',',
 'dua',
 'bulan',
 'bahkan',
 'sudah',
 '1',
 'tahun',
 'lebih',
 'lamanya',
 ',',
 'ayah',
 'Malin',
 'tidak',
 'juga',
 'kembali',
 'ke',
 'kampung',
 'halamannya.Sehingga',
 'ibunya',
 'harus',
 'menggantikan',
 'posisi',
 'ayah',
 'Malin',
 'untuk',
 'mencari',
 'nafkah.Malin',
 'termasuk',
 'anak

In [12]:
hasil_encoding = pd.get_dummies(corpus)
hasil_encoding

!  '  ''  ,  -buahan  -manggil  -masing  -mutar  .  ..  ...  yata  \
0      0  0   0  0        0         0        0       0  0   0  ...     0   
1      0  0   0  0        0         0        0       0  0   0  ...     0   
2      0  0   0  0        0         0        0       0  0   0  ...     0   
3      0  0   0  0        0         0        0       0  0   0  ...     0   
4      0  0   0  0        0         0        0       0  0   0  ...     0   
...   .. ..  .. ..      ...       ...      ...     ... ..  ..  ...   ...   
39398  0  0   0  1        0         0        0       0  0   0  ...     0   
39399  0  0   0  0        0         0        0       0  0   0  ...     0   
39400  0  0   0  0        0         0        0       0  0   0  ...     0   
39401  0  0   0  0        0         0        0       0  0   0  ...     0   
39402  0  0   0  0        0         0        0       0  1   0  ...     0   

       yatim  yelamatkan  yet.Mereka  yik  yikan  yir  yuk  yut  zam  
0          0           0           0    0      0    0    0    0    0  
1          0           0           0    0      0    0    0    0    0  
2          0           0           0    0      0    0    0    0    0  
3          0           0           0    0      0    0    0    0    0  
4          0           0           0    0      0    0    0    0    0  
...      ...         ...         ...  ...    ...  ...  ...  ...  ...  
39398      0           0           0    0      0    0    0    0    0  
39399      0           0           0    0      0    0    0    0    0  
39400      0           0           0    0      0    0    0    0    0  
39401      0           0           0    0      0    0    0    0    0  
39402      0           0           0    0      0    0    0    0    0  

[39403 rows x 7629 columns]

In [14]:
hasil_encoding = pd.get_dummies(corpus)
hasil_encoding.head(20)

!  '  ''  ,  -buahan  -manggil  -masing  -mutar  .  ..  ...  yata  yatim  \
0   0  0   0  0        0         0        0       0  0   0  ...     0      0   
1   0  0   0  0        0         0        0       0  0   0  ...     0      0   
2   0  0   0  0        0         0        0       0  0   0  ...     0      0   
3   0  0   0  0        0         0        0       0  0   0  ...     0      0   
4   0  0   0  0        0         0        0       0  0   0  ...     0      0   
5   0  0   0  1        0         0        0       0  0   0  ...     0      0   
6   0  0   0  0        0         0        0       0  0   0  ...     0      0   
7   0  0   0  0        0         0        0       0  0   0  ...     0      0   
8   0  0   0  0        0         0        0       0  0   0  ...     0      0   
9   0  0   0  0        0         0        0       0  0   0  ...     0      0   
10  0  0   0  0        0         0        0       0  0   0  ...     0      0   
11  0  0   0  0        0         0        0       0  0   0  ...     0      0   
12  0  0   0  0        0         0        0       0  0   0  ...     0      0   
13  0  0   0  0        0         0        0       0  0   0  ...     0      0   
14  0  0   0  0        0         0        0       0  0   0  ...     0      0   
15  0  0   0  0        0         0        0       0  0   0  ...     0      0   
16  0  0   0  0        0         0        0       0  0   0  ...     0      0   
17  0  0   0  0        0         0        0       0  0   0  ...     0      0   
18  0  0   0  0        0         0        0       0  0   0  ...     0      0   
19  0  0   0  1        0         0        0       0  0   0  ...     0      0   

    yelamatkan  yet.Mereka  yik  yikan  yir  yuk  yut  zam  
0            0           0    0      0    0    0    0    0  
1            0           0    0      0    0    0    0    0  
2            0           0    0      0    0    0    0    0  
3            0           0    0      0    0    0    0    0  
4            0           0    0      0    0    0    0    0  
5            0           0    0      0    0    0    0    0  
6            0           0    0      0    0    0    0    0  
7            0           0    0      0    0    0    0    0  
8            0           0    0      0    0    0    0    0  
9            0           0    0      0    0    0    0    0  
10           0           0    0      0    0    0    0    0  
11           0           0    0      0    0    0    0    0  
12           0           0    0      0    0    0    0    0  
13           0           0    0      0    0    0    0    0  
14           0           0    0      0    0    0    0    0  
15           0           0    0      0    0    0    0    0  
16           0           0    0      0    0    0    0    0  
17           0           0    0      0    0    0    0    0  
18           0           0    0      0    0    0    0    0  
19           0           0    0      0    0    0    0    0  

[20 rows x 7629 columns]

**Interpretasi Hasil :**

Dari output di atas, kita dapat melihat bahwa pada indeks ke-5 dan indeks ke-19 simbol "," bernilai 1 yang menunjukkan bahwa ada kaitan antara simbol "," dan indeks tersebut atau bahwa simbol "," berada pada index ke-5 dan indeks ke-19 dalam konteks data ini.



#**2. Hash**

In [15]:
from sklearn.feature_extraction.text import HashingVectorizer

vectorizer = HashingVectorizer(n_features=10)
hashed_representation = vectorizer.fit_transform([data])

hash = pd.DataFrame(hashed_representation.toarray())
hash

0         1         2         3         4         5         6  \
0  0.097088 -0.525854 -0.282916  0.027677  0.230199 -0.601855  0.059746   

          7         8         9  
0 -0.075561  0.385275 -0.245574

**Interpretasi Hasil :**

Dari output di atas, kita dapat melihat bahwa Fitur 0, Fitur 3, Fitur 4, Fitur 6, dan Fitur 8 bernilai positif.  Ini artinya bahwa fitur tersebut lebih penting atau lebih relevan dalam cerita tersebut.<P>
Sementara Fitur 1, Fitur 2, Fitur 5, Fitur 7, dan Fitur 9 bernilai negatif. Ini artinya bahwa fitur tersebut kurang relevan dalam cerita tersebut.

#**3. Co-occurance matrix**

In [16]:
vectorizer = CountVectorizer(tokenizer=lambda x: x.split(), ngram_range=(1, 1))
data_occur = vectorizer.fit_transform([data])
hasil = (data_occur.T * data_occur)
hasil.setdiag(0)

# Menyimpan hasil co-occurrence matrix ke dalam DataFrame
occurence = pd.DataFrame(hasil.toarray(), index=vectorizer.get_feature_names_out(), columns=vectorizer.get_feature_names_out())

# Menampilkan DataFrame
occurence

/usr/local/lib/python3.10/dist-packages/sklearn/feature_extraction/text.py:528: UserWarning: The parameter 'token_pattern' will not be used since 'tokenizer' is not None'
  warnings.warn(


!  !"  !",  !".  !,"  !.    "  ",  "a...  "aaa...  ...  yet.  yik  \
!         0  33   44   11   22  11  264  22     11       11  ...    11   11   
!"       33   0   12    3    6   3   72   6      3        3  ...     3    3   
!",      44  12    0    4    8   4   96   8      4        4  ...     4    4   
!".      11   3    4    0    2   1   24   2      1        1  ...     1    1   
!,"      22   6    8    2    0   2   48   4      2        2  ...     2    2   
...      ..  ..  ...  ...  ...  ..  ...  ..    ...      ...  ...   ...  ...   
yosaku,  44  12   16    4    8   4   96   8      4        4  ...     4    4   
yosaku.  55  15   20    5   10   5  120  10      5        5  ...     5    5   
yuk!"    11   3    4    1    2   1   24   2      1        1  ...     1    1   
yut      11   3    4    1    2   1   24   2      1        1  ...     1    1   
zam!     11   3    4    1    2   1   24   2      1        1  ...     1    1   

         yikan  yir  yosaku  yosaku,  yosaku.  yuk!"  yut  zam!  
!           11   22     110       44       55     11   11    11  
!"           3    6      30       12       15      3    3     3  
!",          4    8      40       16       20      4    4     4  
!".          1    2      10        4        5      1    1     1  
!,"          2    4      20        8       10      2    2     2  
...        ...  ...     ...      ...      ...    ...  ...   ...  
yosaku,      4    8      40        0       20      4    4     4  
yosaku.      5   10      50       20        0      5    5     5  
yuk!"        1    2      10        4        5      0    1     1  
yut          1    2      10        4        5      1    0     1  
zam!         1    2      10        4        5      1    1     0  

[7647 rows x 7647 columns]

**Interpretasi Hasil :**

Dari output di atas, kita dapat melihat bahwa pada baris dan kolom dengan label !:<P>
 - Nilai 0 menunjukkan bahwa label ! tidak muncul bersama dirinya sendiri. <P>
 - Nilai 33 menunjukkan bahwa abel ! dan label !" muncul bersama sebanyak 33 kali<P>
 - Nilai 44 menunjukkan bahwa label ! dan label !", muncul bersama sebanyak 44 kali, dan seterusnya. <P>
 - Nilai 264 menunjukkan bahwa label ! dan label " muncul bersama sebanyak 264 kali.

#**4.Word2Vec**

In [21]:
# Tokenisasi teks
sentences = sent_tokenize(data)
tokens = [word_tokenize(sentence.lower()) for sentence in sentences]

# Membangun model Word2Vec
model_word = Word2Vec(tokens, vector_size=10, window=3, min_count=1, workers=4)

# Mendapatkan representasi vektor kata
word_vectors = pd.DataFrame(model_word.wv.vectors, index=model_word.wv.index_to_key)

# Menampilkan top 5 kata dengan vektor tertinggi
top_words = word_vectors.sum(axis=1).nlargest(5).index
top_vectors = word_vectors.loc[top_words]

print("Top 5 kata dengan vektor tertinggi:")
top_vectors

Top 5 kata dengan vektor tertinggi:


0         1         2         3         4         5         6  \
dan     0.429273 -1.572258  3.791660 -1.336716  0.881877  1.908241  2.557869   
untuk   0.376450 -1.406510  3.362652 -1.206781  0.890490  1.879159  2.386142   
yang    0.559781 -1.371267  3.376989 -1.185464  0.899258  1.721372  2.197297   
dengan  0.474840 -1.390382  3.305602 -1.096328  0.919583  1.789473  2.131665   
di      0.528850 -1.233750  3.255797 -1.233422  0.914496  1.751459  2.330369   

               7         8         9  
dan     2.134173 -1.521770 -1.468852  
untuk   2.016955 -1.330095 -1.214264  
yang    2.118026 -1.443604 -1.279459  
dengan  1.921974 -1.363727 -1.147155  
di      2.017634 -1.375809 -1.470023

**Interpretasi Hasil :**

Dari output di atas, kita dapat melihat bahwa setiap baris mewakili kata dan vektor representasi kata dalam model Word2Vec. Misalnya, untuk kata "dan", vektor representasinya adalah [0.429273,	-1.572258,	3.791660,	-1.336716,	0.881877,	1.908241,	2.557869,	2.134173,	-1.521770,	-1.468852]. Ini adalah representasi vektor kata "dan" dalam dimensi yang ditentukan oleh model.

#**5.Fast text**

In [23]:
# Membangun model fasttext
model_fast = FastText(tokens, vector_size=10, window=5, min_count=1, workers=4)

# Mendapatkan representasi vektor kata
vektor_kata = pd.DataFrame(model_fast.wv.vectors, index=model_fast.wv.index_to_key)

# Menampilkan top 5 kata dengan vektor tertinggi
top_fasttext = vektor_kata.sum(axis=1).nlargest(10000).index
top_textfast = vektor_kata.loc[top_fasttext]

print("Top 5 kata dengan vektor tertinggi:")
top_textfast

Top 5 kata dengan vektor tertinggi:


0         1         2         3         4         5         6  \
kan   4.767126  0.702192  0.991068  6.977452  1.032237  5.257308  3.966009   
an    4.301075  0.659704  0.866724  6.290715  0.922694  4.806938  3.601147   
meng  4.197643  0.628399  0.839801  6.190696  0.943773  4.712602  3.553279   
men   4.069458  0.652407  0.840823  6.040432  0.876941  4.605921  3.450023   
tan   3.728569  0.569107  0.773324  5.467711  0.785974  4.155995  3.118276   
...        ...       ...       ...       ...       ...       ...       ...   
'    -0.010100 -0.015595  0.008694  0.029406 -0.027948 -0.071806  0.015988   
5    -0.015684 -0.077377  0.039513  0.033872 -0.033486  0.059152  0.002632   
12    0.017489  0.014432 -0.043494 -0.025777 -0.012417 -0.063046  0.054673   
o    -0.037516 -0.019094  0.003895  0.051331 -0.030917 -0.000705  0.029379   
999  -0.050234 -0.021594 -0.029833  0.025285  0.003900  0.005562 -0.020121   

             7         8         9  
kan   5.643171  1.694575  7.960998  
an    5.133563  1.556964  7.175822  
meng  4.993054  1.558219  7.001636  
men   4.838703  1.516555  6.893784  
tan   4.510615  1.341453  6.242265  
...        ...       ...       ...  
'     0.052654 -0.014475  0.088068  
5    -0.001983  0.058157 -0.016679  
12    0.005563  0.019865 -0.005367  
o    -0.017646 -0.076984  0.029007  
999   0.001764  0.017848 -0.007706  

[5421 rows x 10 columns]

**Interpretasi Hasil :**

Dari output di atas, kita dapat melihat bahwa lima kata teratas dengan vektor tertinggi adalah "kan," "an," "meng," "men," dan "tan." Untuk masing-masing kata, vektor representasinya terdiri dari 10 nilai numerik yang menggambarkan representasi kata tersebut dalam model FastText.